ibis/backends/tests/test_join.py

```diff
@@ -326,3 +326,23 @@ def test_join_with_trivial_predicate(awards_players, predicate, how, pandas_valu
     result = expr.to_pandas()
 
     assert len(result) == len(expected)
+
+
+@pytest.mark.parametrize(
+    ("how", "nrows"),
+    [
+        param("left", 2, id="left"),
+        param("right", 1, id="right"),
+        param("outer", 2, id="outer"),
+    ],
+)
+def test_outer_join_nullability(con, how, nrows):
+    left = con.tables.win.select(x=lambda t: t.x.type().copy(nullable=False)).filter(
+        lambda t: t.x.isin((1, 2))
+    )
+    right = left.filter(lambda t: t.x == 1)
+    expr = left.join(right, "x", how=how)
+    assert all(typ.nullable for typ in expr.schema().types)
```

Comment:
Should `"x"`'s nullability stay as is?

Reply:
No, I don't think so.  There shouldn't be any non-nullable outputs coming out of a join

Reply:
No, because it's possible for it to be null in the join projection **even if the input column is not null**.

Whether a column from a join project is null depends on both the join type, the predicate **and** the data, the latter of which we can't know statically.

```
D create or replace table t (x BIGINT NOT NULL);
D create or replace table s (x BIGINT NOT NULL);
D insert into t values (1), (2);
D insert into s values (1), (2), (3);
D select t.x as x, s.x as sx from t full join s on t.x=s.x;
┌───────┬───────┐
│   x   │  sx   │
│ int64 │ int64 │
├───────┼───────┤
│     1 │     1 │
│     2 │     2 │
│       │     3 │
└───────┴───────┘
```

Reply:
> if you're joining on "x", and that x isn't nullable, could it become nullable? I'm not familiar with how SQL handles nulls in a join column.

It can. See my example 😅

Reply:
Missed that as I was typing the comment hehe, thanks for the clarification :)

Reply:
The nullability for all of our SQL engines except one (ClickHouse) is a constraint on input to an existing table: you can't insert any NULLs into a column with a `NOT NULL` constraint.

The nullability of a function call or expression in a `SELECT` statement is not constrained in any way. It might be used in the engine for optimization but that's way below ibis's operating level.